File: 프로그래머스/lv2/12924. 숫자의 표현/숫자의 표현.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(int n) {
    int answer = 0;
    vector<int> dp(n + 1, 0);
    dp[1] = 1;
    for (int i = 2; i < n + 1; i++) {
        dp[i] = dp[i - 1] + i;
    }
    // for (int i = 0; i < n + 1; i++) cout << dp[i] << " ";
    // cout << "\n";
    int start = 1, end = 0;
    while (start <= n) {
        // cout << start << ", " << end << " : " << dp[start] - dp[end] << "\n";
        if (dp[start] - dp[end] > n) end++;
        else if (dp[start] - dp[end] < n) start++;
        else {
            answer++;
            start++;
        }
    }
    return answer;
}

File: 프로그래머스/lv2/12924. 숫자의 표현/숫자의 표현.py
def solution(n):
    answer = 0
    dp = [0] * (n + 1)
    for i in range(1, n + 1) : dp[i] += dp[i - 1] + i
    start, end = 1, 0
    while start <= n :
        if dp[start] - dp[end] < n : start += 1
        elif dp[start] - dp[end] > n : end += 1
        else :
            answer += 1
            end += 1
    return answer